pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,420 @@
+		}
+		if props.ArtifactPath.IsAbs() {
+			return nil, errors.New("dsse envelopes cannot be fetched over HTTP(S)")
+		}
+		artifactBytes, err = os.ReadFile(filepath.Clean(props.ArtifactPath.Path))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	env := &dsse.Envelope{}
+	if err := json.Unmarshal(artifactBytes, env); err != nil {
+		return nil, fmt.Errorf("payload must be a valid DSSE envelope: %w", err)
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	if len(props.PublicKeyBytes) > 0 {
+		allPubKeyBytes = append(allPubKeyBytes, props.PublicKeyBytes...)
+	}
+
+	if len(props.PublicKeyPaths) > 0 {
+		for _, path := range props.PublicKeyPaths {
+			if path.IsAbs() {
+				return nil, errors.New("dsse public keys cannot be fetched over HTTP(S)")
+			}
+
+			publicKeyBytes, err := os.ReadFile(filepath.Clean(path.Path))
+			if err != nil {
+				return nil, fmt.Errorf("error reading public key file: %w", err)
+			}
+
+			allPubKeyBytes = append(allPubKeyBytes, publicKeyBytes)
+		}
+	}
+
+	keysBySig, err := verifyEnvelope(allPubKeyBytes, env)
+	if err != nil {
+		return nil, err
+	}
+	for _, key := range keysBySig {
+		canonicalKey, err := key.CanonicalValue()
+		if err != nil {
+			return nil, err
+		}
+		re.DSSEObj.ProposedContent.PublicKeys = append(re.DSSEObj.ProposedContent.PublicKeys, strfmt.Base64(canonicalKey))
+	}
+	re.DSSEObj.ProposedContent.Envelope = swag.String(string(artifactBytes))
+
+	returnVal.Spec = re.DSSEObj
+	returnVal.APIVersion = swag.String(re.APIVersion())
+
+	return &returnVal, nil
+}
+
+// verifyEnvelope takes in an array of possible key bytes and attempts to parse them as x509 public keys.
+// it then uses these to verify the envelope and makes sure that every signature on the envelope is verified.
+// it returns a map of verifiers indexed by the signature the verifier corresponds to.
+func verifyEnvelope(allPubKeyBytes [][]byte, env *dsse.Envelope) (map[string]*x509.PublicKey, error) {
+	// generate a fake id for these keys so we can get back to the key bytes and match them to their corresponding signature
+	verifierBySig := make(map[string]*x509.PublicKey)
+	allSigs := make(map[string]struct{})
+	for _, sig := range env.Signatures {
+		allSigs[sig.Sig] = struct{}{}
+	}
+
+	for _, pubKeyBytes := range allPubKeyBytes {
+		if len(allSigs) == 0 {
+			break // if all signatures have been verified, do not attempt anymore
+		}
+		key, err := x509.NewPublicKey(bytes.NewReader(pubKeyBytes))
+		if err != nil {
+			return nil, fmt.Errorf("could not parse public key as x509: %w", err)
+		}
+
+		vfr, err := signature.LoadVerifier(key.CryptoPubKey(), crypto.SHA256)
+		if err != nil {
+			return nil, fmt.Errorf("could not load verifier: %w", err)
+		}
+
+		dsseVfr, err := dsse.NewEnvelopeVerifier(&sigdsse.VerifierAdapter{SignatureVerifier: vfr})
+		if err != nil {
+			return nil, fmt.Errorf("could not use public key as a dsse verifier: %w", err)
+		}
+
+		accepted, err := dsseVfr.Verify(context.Background(), env)
```

Comment:
should you pass the context from the caller? CreateFromArtifactProps has a context, but Unmarshal doesnt

Reply:
I'd either have to set the request context as a variable on the struct as otherwise I'd have to make a change to the golang interface

Reply:
looking at this, it will just be an in-memory RSA/ECDSA/ED25519 verification so context isn't used anywhere.